<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>RNA-seq analysis in R</title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="Software Carpentry Blog" href="http://software-carpentry.org/feed.xml"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <a href="index.html"><h1 class="title">RNA-seq analysis in R</h1></a>
          <h2 class="subtitle">Solutions: Analysis of Pasilla Knock-down Experiment in Drosophila</h2>
          <p><strong>Author: Belinda Phipson and Jovana Maksimovic</strong></p>
<h2 id="data-files-needed">Data files needed</h2>
<ul>
<li>counts_Drosophila.txt</li>
<li>targetsD.txt</li>
</ul>
<p>Available from <a href="https://figshare.com/s/e08e71c42f118dbe8be6" class="uri">https://figshare.com/s/e08e71c42f118dbe8be6</a>.</p>
<h2 id="libraries-needed">Libraries needed</h2>
<ul>
<li>limma</li>
<li>edgeR</li>
<li>org.Dm.eg.db</li>
<li>EDASeq</li>
<li>gplots</li>
<li>RColorBrewer</li>
</ul>
<h2 id="introduction">Introduction</h2>
<p>The RNA-Seq data we will be analysing today come from this published paper:</p>
<p>Brooks, A.N., Yang, L., Duff, M.O., Hansen, K.D., Park, J.W., Dudoit, S., Brenner, S.E. and Graveley, B.R. (2011) Conservation of an rna regulatory map between drosophila and mammals. Genome Research, 21(2), 193-202.<br />
http://www.ncbi.nlm.nih.gov/pubmed/20921232</p>
<p>This is a publicly available dataset, deposited in the Short Read Archive. The RNA-sequence data are available from GEO under accession nos. GSM461176-GSM461181. The authors combined RNAi and RNASeq to identify exons regulated by Pasilla, the Drosophila melanogaster ortholog of mammalian NOVA1 and NOVA2. They showed that the RNA regulatory map of Pasilla and NOVA1/2 is highly conserved between insects and mammals. NOVA1 and NOVA2 are best known for being involved in alternative splicing. Cells from S2-DRSC, which is an embryonic cell line, were cultured and subjected to a treatment in order to knock-down Pasilla. The four untreated and three treated RNAi samples were used in the analysis. The treated samples had Pasilla knocked down by approximately 60% compared to the untreated samples. Some of the samples had undergone paired end sequencing while other samples were sequenced from one end only.</p>
<p>The reads were aligned to the Drosophila reference genome, downloaded from Ensembl, using the tophat aligner. The reads were summarised at the gene-level using htseq-count, a function from the tool HTSeq (http://wwwhuber.embl.de/users/anders/HTSeq/doc/overview.html).</p>
<p>For the purpose of today’s workshop, we will be analysing the gene level counts.</p>
<h2 id="reading-data-into-r">Reading data into R</h2>
<p>First, let’s load all the libraries we will need today.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(limma)
<span class="kw">library</span>(edgeR)
<span class="kw">library</span>(EDASeq)
<span class="kw">library</span>(gplots)
<span class="kw">library</span>(RColorBrewer)
<span class="kw">library</span>(org.Dm.eg.db)</code></pre></div>
<p>Next, read in the data and targets file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="dt">file=</span><span class="st">&quot;data/counts_Drosophila.txt&quot;</span>)
targets &lt;-<span class="st"> </span><span class="kw">read.delim</span>(<span class="dt">file=</span><span class="st">&quot;data/SampleInfo_Drosophila.txt&quot;</span>)</code></pre></div>
<p>Check that the data has read in correctly</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(counts)</code></pre></div>
<pre class="output"><code>            SRR031714.bam SRR031716.bam SRR031724.bam SRR031726.bam
FBgn0037213           157           142           213           291
FBgn0000500             0             3             4             5
FBgn0053294            14            18            13            19
FBgn0037215          1666          1948          1662          1833
FBgn0037217            13            19            18            23
FBgn0037218           732           755           803           906
            SRR031708.bam SRR031718.bam SRR031728.bam
FBgn0037213           123           225           164
FBgn0000500             2             1             1
FBgn0053294            22            34            22
FBgn0037215          1625          1701          1870
FBgn0037217            15            15            17
FBgn0037218           695           767           793
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">targets</code></pre></div>
<pre class="output"><code>  SampleName     Group Library
1  SRR031714 Untreated      PE
2  SRR031716 Untreated      PE
3  SRR031724   Treated      PE
4  SRR031726   Treated      PE
5  SRR031708 Untreated      SE
6  SRR031718   Treated      SE
7  SRR031728 Untreated      SE
</code></pre>
<h2 id="filtering-out-lowly-expressed-genes">Filtering out lowly expressed genes</h2>
<p>Our main interest is in testing the treated versus untreated groups. To check how many samples we have in each group we can use the <code>table</code> command.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(targets$Group)</code></pre></div>
<pre class="output"><code>
  Treated Untreated 
        3         4 
</code></pre>
<p>The minimum sample size is 3. Let’s check the relationship between CPM and counts to see what CPM threshold we should be imposing. Recall we’re looking for a CPM that corresponds to a count of roughly 10-15.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mycpm &lt;-<span class="st"> </span><span class="kw">cpm</span>(counts)
<span class="kw">plot</span>(counts[,<span class="dv">1</span>],mycpm[,<span class="dv">1</span>],<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">20</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">5</span>))
<span class="kw">abline</span>(<span class="dt">v=</span><span class="dv">10</span>,<span class="dt">col=</span><span class="dv">2</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">2</span>,<span class="dt">col=</span><span class="dv">4</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-6-1.png" title="plot of chunk unnamed-chunk-6" alt="plot of chunk unnamed-chunk-6" style="display: block; margin: auto;" /></p>
<p>We can filter on a CPM of 2 or 3 in at least 3 samples.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">thresh &lt;-<span class="st"> </span>mycpm &gt;<span class="st"> </span><span class="dv">2</span>
keep &lt;-<span class="st"> </span><span class="kw">rowSums</span>(thresh) &gt;=<span class="st"> </span><span class="dv">3</span>
<span class="kw">table</span>(keep)</code></pre></div>
<pre class="output"><code>keep
FALSE  TRUE 
 7345  7524 
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">counts.keep &lt;-<span class="st"> </span>counts[keep,]
<span class="kw">dim</span>(counts.keep)</code></pre></div>
<pre class="output"><code>[1] 7524    7
</code></pre>
<p>We are filtering out about half our genes.</p>
<h2 id="convert-to-dgelist-object">Convert to DGEList object</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">DGEList</span>(counts.keep)</code></pre></div>
<h2 id="quality-control">Quality control</h2>
<p>Let’s do a number of quality control plots.</p>
<p>First, check the library sizes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">barplot</span>(y$samples$lib.size)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-9-1.png" title="plot of chunk unnamed-chunk-9" alt="plot of chunk unnamed-chunk-9" style="display: block; margin: auto;" /></p>
<p>Next check the distribution of the counts using a boxplot:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="co"># Get log2 counts per million</span>
logcpm &lt;-<span class="st"> </span><span class="kw">cpm</span>(y$counts,<span class="dt">log=</span><span class="ot">TRUE</span>)
<span class="co"># Check distributions of samples using boxplots</span>
<span class="kw">boxplot</span>(logcpm, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Log2 counts per million&quot;</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">outline=</span><span class="ot">FALSE</span>)
<span class="co"># Let&#39;s add a blue horizontal line that corresponds to the median logCPM</span>
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(logcpm),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)
<span class="kw">title</span>(<span class="st">&quot;Boxplots of logCPMs (unnormalised)&quot;</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-10-1.png" title="plot of chunk unnamed-chunk-10" alt="plot of chunk unnamed-chunk-10" style="display: block; margin: auto;" /></p>
<p>We can colour by our groups, or by the different library prep.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>),<span class="dt">oma=</span><span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))
group.col &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;blue&quot;</span>)[targets$Group]
<span class="kw">boxplot</span>(logcpm, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Log2 counts per million&quot;</span>,<span class="dt">las=</span><span class="dv">2</span>,<span class="dt">col=</span>group.col,
        <span class="dt">pars=</span><span class="kw">list</span>(<span class="dt">cex.lab=</span><span class="fl">0.8</span>,<span class="dt">cex.axis=</span><span class="fl">0.8</span>))
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(logcpm),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)
<span class="kw">title</span>(<span class="st">&quot;Boxplots of logCPMs</span><span class="ch">\n</span><span class="st">(coloured by groups)&quot;</span>,<span class="dt">cex.main=</span><span class="fl">0.8</span>)

lib.col &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;light pink&quot;</span>,<span class="st">&quot;light green&quot;</span>)[targets$Library]
<span class="kw">boxplot</span>(logcpm, <span class="dt">xlab=</span><span class="st">&quot;&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;Log2 counts per million&quot;</span>,<span class="dt">las=</span><span class="dv">2</span>, <span class="dt">col=</span>lib.col,
        <span class="dt">pars=</span><span class="kw">list</span>(<span class="dt">cex.lab=</span><span class="fl">0.8</span>,<span class="dt">cex.axis=</span><span class="fl">0.8</span>))
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(logcpm),<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)
<span class="kw">title</span>(<span class="st">&quot;Boxplots of logCPMs</span><span class="ch">\n</span><span class="st">(coloured by library prep)&quot;</span>,<span class="dt">cex.main=</span><span class="fl">0.8</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-11-1.png" title="plot of chunk unnamed-chunk-11" alt="plot of chunk unnamed-chunk-11" style="display: block; margin: auto;" /></p>
<p>It doesn’t look like there are any large biases in the data.</p>
<p>Finally, let’s check our MDS plots.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">plotMDS</span>(y)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-12-1.png" title="plot of chunk unnamed-chunk-12" alt="plot of chunk unnamed-chunk-12" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMDS</span>(y,<span class="dt">col=</span>group.col)
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>,<span class="dt">legend=</span><span class="kw">levels</span>(targets$Group),<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>,<span class="st">&quot;blue&quot;</span>))
<span class="kw">plotMDS</span>(y,<span class="dt">col=</span>lib.col)
<span class="kw">legend</span>(<span class="st">&quot;topleft&quot;</span>,<span class="dt">legend=</span><span class="kw">levels</span>(targets$Library),<span class="dt">fill=</span><span class="kw">c</span>(<span class="st">&quot;light pink&quot;</span>,<span class="st">&quot;light green&quot;</span>))</code></pre></div>
<p><img src="fig-09/unnamed-chunk-13-1.png" title="plot of chunk unnamed-chunk-13" alt="plot of chunk unnamed-chunk-13" style="display: block; margin: auto;" /></p>
<p>It looks like there is some variability in the data due to library type e.g. single end or paired end.</p>
<h2 id="hierarchical-clustering-with-heatmap.2">Hierarchical clustering with heatmap.2</h2>
<p>First we need a matrix of log counts:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">logcounts &lt;-<span class="st"> </span><span class="kw">cpm</span>(y,<span class="dt">log=</span><span class="ot">TRUE</span>)</code></pre></div>
<p>Get variances for genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">var_genes &lt;-<span class="st"> </span><span class="kw">apply</span>(logcounts, <span class="dv">1</span>, var)</code></pre></div>
<p>Get top 500 most variable</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">select_var &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(var_genes, <span class="dt">decreasing=</span><span class="ot">TRUE</span>))[<span class="dv">1</span>:<span class="dv">500</span>]</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">highly_variable_lcpm &lt;-<span class="st"> </span>logcounts[select_var,]
<span class="kw">dim</span>(highly_variable_lcpm)</code></pre></div>
<pre class="output"><code>[1] 500   7
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mypalette &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">11</span>,<span class="st">&quot;RdYlBu&quot;</span>)
morecols &lt;-<span class="st"> </span><span class="kw">colorRampPalette</span>(mypalette)
<span class="co"># Plot the heatmap</span>
<span class="kw">heatmap.2</span>(highly_variable_lcpm,<span class="dt">col=</span><span class="kw">rev</span>(<span class="kw">morecols</span>(<span class="dv">50</span>)),<span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">main=</span><span class="st">&quot;Top 500 most variable genes across samples&quot;</span>,<span class="dt">ColSideColors=</span>group.col,<span class="dt">scale=</span><span class="st">&quot;row&quot;</span>,<span class="dt">margins=</span><span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">5</span>))</code></pre></div>
<p><img src="fig-09/unnamed-chunk-18-1.png" title="plot of chunk unnamed-chunk-18" alt="plot of chunk unnamed-chunk-18" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">heatmap.2</span>(highly_variable_lcpm,<span class="dt">col=</span><span class="kw">rev</span>(<span class="kw">morecols</span>(<span class="dv">50</span>)),<span class="dt">trace=</span><span class="st">&quot;none&quot;</span>, <span class="dt">main=</span><span class="st">&quot;Top 500 most variable genes across samples&quot;</span>,<span class="dt">ColSideColors=</span>lib.col,<span class="dt">scale=</span><span class="st">&quot;row&quot;</span>,<span class="dt">margins=</span><span class="kw">c</span>(<span class="dv">10</span>,<span class="dv">5</span>))</code></pre></div>
<p><img src="fig-09/unnamed-chunk-18-2.png" title="plot of chunk unnamed-chunk-18" alt="plot of chunk unnamed-chunk-18" style="display: block; margin: auto;" /></p>
<h2 id="normalisation">Normalisation</h2>
<p>Let’s do TMM normalisation</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y &lt;-<span class="st"> </span><span class="kw">calcNormFactors</span>(y)
y$samples</code></pre></div>
<pre class="output"><code>              group lib.size norm.factors
SRR031714.bam     1  4198181    0.9805029
SRR031716.bam     1  4740969    0.9566649
SRR031724.bam     1  4601291    0.9936278
SRR031726.bam     1  5195516    1.0019699
SRR031708.bam     1  3980986    1.0036556
SRR031718.bam     1  4320527    1.0556860
SRR031728.bam     1  4031608    1.0106328
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMD</span>(logcounts,<span class="dt">column=</span><span class="dv">2</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)
<span class="kw">plotMD</span>(y,<span class="dt">column =</span> <span class="dv">2</span>)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-20-1.png" title="plot of chunk unnamed-chunk-20" alt="plot of chunk unnamed-chunk-20" style="display: block; margin: auto;" /></p>
<h2 id="differential-expression">Differential expression</h2>
<h3 id="set-up-design-matrix">Set up design matrix</h3>
<p>We want to test for differences between the treated and untreated samples. However, we know that the library preparation adds variability to the data, so we need to account for it in our model. We do this by modelling both Group and Library as variables in our design matrix. This is known as an additive model.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">design &lt;-<span class="st"> </span><span class="kw">model.matrix</span>(~targets$Library +<span class="st"> </span>targets$Group)
design</code></pre></div>
<pre class="output"><code>  (Intercept) targets$LibrarySE targets$GroupUntreated
1           1                 0                      1
2           1                 0                      1
3           1                 0                      0
4           1                 0                      0
5           1                 1                      1
6           1                 1                      0
7           1                 1                      1
attr(,&quot;assign&quot;)
[1] 0 1 2
attr(,&quot;contrasts&quot;)
attr(,&quot;contrasts&quot;)$`targets$Library`
[1] &quot;contr.treatment&quot;

attr(,&quot;contrasts&quot;)$`targets$Group`
[1] &quot;contr.treatment&quot;
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">colnames</span>(design) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Int&quot;</span>,<span class="st">&quot;SEvsPE&quot;</span>,<span class="st">&quot;UVsT&quot;</span>)</code></pre></div>
<h3 id="voom-transform-the-data">Voom transform the data</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
v &lt;-<span class="st"> </span><span class="kw">voom</span>(y,design,<span class="dt">plot=</span><span class="ot">TRUE</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-22-1.png" title="plot of chunk unnamed-chunk-22" alt="plot of chunk unnamed-chunk-22" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">boxplot</span>(logcounts)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(logcounts),<span class="dt">col=</span><span class="dv">4</span>)
<span class="kw">boxplot</span>(v$E)
<span class="kw">abline</span>(<span class="dt">h=</span><span class="kw">median</span>(v$E),<span class="dt">col=</span><span class="dv">4</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-23-1.png" title="plot of chunk unnamed-chunk-23" alt="plot of chunk unnamed-chunk-23" style="display: block; margin: auto;" /></p>
<h3 id="test-for-differential-expression">Test for differential expression</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit &lt;-<span class="st"> </span><span class="kw">lmFit</span>(v,design)
fit &lt;-<span class="st"> </span><span class="kw">eBayes</span>(fit)
results &lt;-<span class="st"> </span><span class="kw">decideTests</span>(fit)
<span class="kw">summary</span>(results)</code></pre></div>
<pre class="output"><code>    Int SEvsPE UVsT
-1    4    125  440
0    66   7142 6641
1  7454    257  443
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit,<span class="dt">coef=</span><span class="dv">3</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>                logFC   AveExpr         t      P.Value    adj.P.Val
FBgn0025111 -2.863110  6.253448 -31.24326 9.756044e-13 4.734428e-09
FBgn0003360  3.089242  7.778378  30.57062 1.258487e-12 4.734428e-09
FBgn0026562  2.422219 11.359837  24.25592 1.868075e-11 4.685133e-08
FBgn0029167  2.247086  7.825170  21.77945 6.509505e-11 1.224438e-07
FBgn0035085  2.677008  5.180607  18.97509 3.188148e-10 4.797524e-07
FBgn0039155  4.272254  4.800907  17.42222 8.476440e-10 1.062946e-06
FBgn0040091  1.559497  6.245766  15.43293 3.363833e-09 3.615639e-06
FBgn0023479  1.576335  7.768886  15.06473 4.419245e-09 3.877495e-06
FBgn0029896  2.312097  4.881577  15.00033 4.638152e-09 3.877495e-06
FBgn0027279  1.180272  7.760889  14.37303 7.503535e-09 5.645660e-06
                   B
FBgn0025111 19.04663
FBgn0003360 19.04370
FBgn0026562 16.57869
FBgn0029167 15.57725
FBgn0035085 13.66077
FBgn0039155 12.15024
FBgn0040091 11.72125
FBgn0023479 11.43434
FBgn0029896 11.25267
FBgn0027279 10.89435
</code></pre>
<h3 id="add-annotation-from-org.dm.eg.db">Add annotation from org.Dm.eg.db</h3>
<p>The rownames of the fit object are <a href="FlyBase">http://flybase.org/</a> ids.</p>
<p>First we need to decide what information we want. In order to see what we can extract we can run the <code>columns</code> function on the annotation database.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">columns</span>(org.Dm.eg.db)</code></pre></div>
<pre class="output"><code> [1] &quot;ACCNUM&quot;       &quot;ALIAS&quot;        &quot;ENSEMBL&quot;      &quot;ENSEMBLPROT&quot; 
 [5] &quot;ENSEMBLTRANS&quot; &quot;ENTREZID&quot;     &quot;ENZYME&quot;       &quot;EVIDENCE&quot;    
 [9] &quot;EVIDENCEALL&quot;  &quot;FLYBASE&quot;      &quot;FLYBASECG&quot;    &quot;FLYBASEPROT&quot; 
[13] &quot;GENENAME&quot;     &quot;GO&quot;           &quot;GOALL&quot;        &quot;MAP&quot;         
[17] &quot;ONTOLOGY&quot;     &quot;ONTOLOGYALL&quot;  &quot;PATH&quot;         &quot;PMID&quot;        
[21] &quot;REFSEQ&quot;       &quot;SYMBOL&quot;       &quot;UNIGENE&quot;      &quot;UNIPROT&quot;     
</code></pre>
<p>We definitely want gene symbols and perhaps the full gene name and Entrez id. Let’s build up our annotation information in a separate data frame using the <code>select</code> function. Note, by default, the <code>select</code> function assumes that the keys provided are Entrez ids. However, in this case we are using FlyBase ids as the keys. As such, we need to give the <code>select</code> function this information using the <code>keytype</code> argument which we will set to <code>&quot;FLYBASE&quot;</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ann &lt;-<span class="st"> </span><span class="kw">select</span>(org.Dm.eg.db,<span class="dt">keys=</span><span class="kw">rownames</span>(fit),<span class="dt">columns=</span><span class="kw">c</span>(<span class="st">&quot;FLYBASE&quot;</span>,<span class="st">&quot;ENTREZID&quot;</span>,<span class="st">&quot;SYMBOL&quot;</span>,<span class="st">&quot;GENENAME&quot;</span>),<span class="dt">keytype=</span><span class="st">&quot;FLYBASE&quot;</span>)</code></pre></div>
<pre class="output"><code>&#39;select()&#39; returned 1:1 mapping between keys and columns
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Have a look at the annotation</span>
<span class="kw">head</span>(ann)</code></pre></div>
<pre class="output"><code>      FLYBASE ENTREZID   SYMBOL
1 FBgn0037213    40522  CG12581
2 FBgn0053294  3772637  CR33294
3 FBgn0037215    40524 beta-Man
4 FBgn0037217    40526  CG14636
5 FBgn0037218    40527      aux
6 FBgn0261436    40528     DhpD
                                         GENENAME
1 CG12581 gene product from transcript CG12581-RB
2                                          pseudo
3                                beta-Mannosidase
4 CG14636 gene product from transcript CG14636-RA
5                                         auxilin
6                         Dihydropterin deaminase
</code></pre>
<p>Let’s double check that the <code>FLYBASE</code> column matches exactly to our <code>fit</code> rownames.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">table</span>(ann$FLYBASE==<span class="kw">rownames</span>(fit))</code></pre></div>
<pre class="output"><code>
TRUE 
7524 
</code></pre>
<p>We can slot in the annotation information into the <code>genes</code> slot of <code>fit</code>. (Please note that if the <code>select</code> function returns a 1:many mapping then you can’t just append the annotation to the fit object. An alternative way to get annotation is shown below.)</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit$genes &lt;-<span class="st"> </span>ann</code></pre></div>
<p>Now when we run the <code>topTable</code> command, the annotation information should be included in the output.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit,<span class="dt">coef=</span><span class="dv">3</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>                FLYBASE ENTREZID    SYMBOL
FBgn0025111 FBgn0025111    32008      Ant2
FBgn0003360 FBgn0003360    32007      sesB
FBgn0026562 FBgn0026562    43230     SPARC
FBgn0029167 FBgn0029167    39529       Hml
FBgn0035085 FBgn0035085    37991    CG3770
FBgn0039155 FBgn0039155    42865      Kal1
FBgn0040091 FBgn0040091    37590   Ugt58Fa
FBgn0023479 FBgn0023479    39048   Tequila
FBgn0029896 FBgn0029896    31612    CG3168
FBgn0027279 FBgn0027279    33137 l(1)G0196
                                                 GENENAME     logFC
FBgn0025111              Adenine nucleotide translocase 2 -2.863110
FBgn0003360                            stress-sensitive B  3.089242
FBgn0026562       Secreted protein, acidic, cysteine-rich  2.422219
FBgn0029167                                    Hemolectin  2.247086
FBgn0035085 CG3770 gene product from transcript CG3770-RA  2.677008
FBgn0039155                  Kallmann syndrome 1 ortholog  4.272254
FBgn0040091 CG4414 gene product from transcript CG4414-RB  1.559497
FBgn0023479 CG4821 gene product from transcript CG4821-RG  1.576335
FBgn0029896 CG3168 gene product from transcript CG3168-RG  2.312097
FBgn0027279                              lethal (1) G0196  1.180272
              AveExpr         t      P.Value    adj.P.Val        B
FBgn0025111  6.253448 -31.24326 9.756044e-13 4.734428e-09 19.04663
FBgn0003360  7.778378  30.57062 1.258487e-12 4.734428e-09 19.04370
FBgn0026562 11.359837  24.25592 1.868075e-11 4.685133e-08 16.57869
FBgn0029167  7.825170  21.77945 6.509505e-11 1.224438e-07 15.57725
FBgn0035085  5.180607  18.97509 3.188148e-10 4.797524e-07 13.66077
FBgn0039155  4.800907  17.42222 8.476440e-10 1.062946e-06 12.15024
FBgn0040091  6.245766  15.43293 3.363833e-09 3.615639e-06 11.72125
FBgn0023479  7.768886  15.06473 4.419245e-09 3.877495e-06 11.43434
FBgn0029896  4.881577  15.00033 4.638152e-09 3.877495e-06 11.25267
FBgn0027279  7.760889  14.37303 7.503535e-09 5.645660e-06 10.89435
</code></pre>
<p>If for some reason the <code>select</code> function doesn’t work, or the result is a 1:many mapping, there is a less elegant way to get the annotations using the <code>toTable</code> command.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Let&#39;s see what we can get in table format from org.Dm.eg.db</span>
<span class="kw">ls</span>(<span class="st">&quot;package:org.Dm.eg.db&quot;</span>)</code></pre></div>
<pre class="output"><code> [1] &quot;org.Dm.eg&quot;                &quot;org.Dm.eg_dbconn&quot;        
 [3] &quot;org.Dm.eg_dbfile&quot;         &quot;org.Dm.eg_dbInfo&quot;        
 [5] &quot;org.Dm.eg_dbschema&quot;       &quot;org.Dm.eg.db&quot;            
 [7] &quot;org.Dm.egACCNUM&quot;          &quot;org.Dm.egACCNUM2EG&quot;      
 [9] &quot;org.Dm.egALIAS2EG&quot;        &quot;org.Dm.egCHR&quot;            
[11] &quot;org.Dm.egCHRLENGTHS&quot;      &quot;org.Dm.egCHRLOC&quot;         
[13] &quot;org.Dm.egCHRLOCEND&quot;       &quot;org.Dm.egENSEMBL&quot;        
[15] &quot;org.Dm.egENSEMBL2EG&quot;      &quot;org.Dm.egENSEMBLPROT&quot;    
[17] &quot;org.Dm.egENSEMBLPROT2EG&quot;  &quot;org.Dm.egENSEMBLTRANS&quot;   
[19] &quot;org.Dm.egENSEMBLTRANS2EG&quot; &quot;org.Dm.egENZYME&quot;         
[21] &quot;org.Dm.egENZYME2EG&quot;       &quot;org.Dm.egFLYBASE&quot;        
[23] &quot;org.Dm.egFLYBASE2EG&quot;      &quot;org.Dm.egFLYBASECG&quot;      
[25] &quot;org.Dm.egFLYBASECG2EG&quot;    &quot;org.Dm.egFLYBASEPROT&quot;    
[27] &quot;org.Dm.egFLYBASEPROT2EG&quot;  &quot;org.Dm.egGENENAME&quot;       
[29] &quot;org.Dm.egGO&quot;              &quot;org.Dm.egGO2ALLEGS&quot;      
[31] &quot;org.Dm.egGO2EG&quot;           &quot;org.Dm.egMAP&quot;            
[33] &quot;org.Dm.egMAP2EG&quot;          &quot;org.Dm.egMAPCOUNTS&quot;      
[35] &quot;org.Dm.egORGANISM&quot;        &quot;org.Dm.egPATH&quot;           
[37] &quot;org.Dm.egPATH2EG&quot;         &quot;org.Dm.egPMID&quot;           
[39] &quot;org.Dm.egPMID2EG&quot;         &quot;org.Dm.egREFSEQ&quot;         
[41] &quot;org.Dm.egREFSEQ2EG&quot;       &quot;org.Dm.egSYMBOL&quot;         
[43] &quot;org.Dm.egSYMBOL2EG&quot;       &quot;org.Dm.egUNIGENE&quot;        
[45] &quot;org.Dm.egUNIGENE2EG&quot;      &quot;org.Dm.egUNIPROT&quot;        
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Get annotation</span>
fly &lt;-<span class="st"> </span><span class="kw">toTable</span>(org.Dm.egFLYBASE)
<span class="kw">head</span>(fly)</code></pre></div>
<pre class="output"><code>  gene_id  flybase_id
1   30970 FBgn0040373
2   30971 FBgn0040372
3   30972 FBgn0261446
4   30973 FBgn0000316
5   30975 FBgn0005427
6   30976 FBgn0040370
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">symbol &lt;-<span class="st"> </span><span class="kw">toTable</span>(org.Dm.egSYMBOL)
genename &lt;-<span class="st"> </span><span class="kw">toTable</span>(org.Dm.egGENENAME)

<span class="co"># We can use the merge command to merge two dataframes</span>
ann1 &lt;-<span class="st"> </span><span class="kw">merge</span>(fly,symbol,<span class="dt">by=</span><span class="st">&quot;gene_id&quot;</span>)
<span class="kw">head</span>(ann1)</code></pre></div>
<pre class="output"><code>   gene_id  flybase_id  symbol
1 10178776 FBgn0261702 CR42738
2 10178777 FBgn0262024 CG42835
3 10178779 FBgn0058354 CR40354
4 10178780 FBgn0053929 CR33929
5 10178781 FBgn0262141 CG42867
6 10178782 FBgn0062978 CG31808
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Add genename table to ann1</span>
ann2 &lt;-<span class="st"> </span><span class="kw">merge</span>(ann1,genename,<span class="dt">by=</span><span class="st">&quot;gene_id&quot;</span>)
<span class="kw">head</span>(ann2)</code></pre></div>
<pre class="output"><code>   gene_id  flybase_id  symbol
1 10178776 FBgn0261702 CR42738
2 10178777 FBgn0262024 CG42835
3 10178779 FBgn0058354 CR40354
4 10178780 FBgn0053929 CR33929
5 10178781 FBgn0262141 CG42867
6 10178782 FBgn0062978 CG31808
                                        gene_name
1                                           ncRNA
2 CG42835 gene product from transcript CG42835-RA
3                                          pseudo
4                                          pseudo
5 CG42867 gene product from transcript CG42867-RB
6 CG31808 gene product from transcript CG31808-RC
</code></pre>
<p>Now we need to match up between <code>rownames(fit)</code> and the ensemble gene id in <code>ann2</code>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">m &lt;-<span class="st"> </span><span class="kw">match</span>(<span class="kw">rownames</span>(fit),ann2$flybase_id)
<span class="kw">table</span>(<span class="kw">is.na</span>(m)) <span class="co"># check for unmatched rows</span></code></pre></div>
<pre class="output"><code>
FALSE  TRUE 
 7010   514 
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ann3 &lt;-<span class="st"> </span>ann2[m[!<span class="kw">is.na</span>(m)],] <span class="co"># exclude unmatched rows</span>

<span class="co"># compare the results in ann3 to what is in the fit object</span>
<span class="kw">head</span>(ann3)</code></pre></div>
<pre class="output"><code>      gene_id  flybase_id   symbol
16166   40522 FBgn0037213  CG12581
13702 3772637 FBgn0053294  CR33294
16167   40524 FBgn0037215 beta-Man
16168   40526 FBgn0037217  CG14636
16169   40527 FBgn0037218      aux
16170   40528 FBgn0261436     DhpD
                                            gene_name
16166 CG12581 gene product from transcript CG12581-RB
13702                                          pseudo
16167                                beta-Mannosidase
16168 CG14636 gene product from transcript CG14636-RA
16169                                         auxilin
16170                         Dihydropterin deaminase
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(fit$genes)</code></pre></div>
<pre class="output"><code>      FLYBASE ENTREZID   SYMBOL
1 FBgn0037213    40522  CG12581
2 FBgn0053294  3772637  CR33294
3 FBgn0037215    40524 beta-Man
4 FBgn0037217    40526  CG14636
5 FBgn0037218    40527      aux
6 FBgn0261436    40528     DhpD
                                         GENENAME
1 CG12581 gene product from transcript CG12581-RB
2                                          pseudo
3                                beta-Mannosidase
4 CG14636 gene product from transcript CG14636-RA
5                                         auxilin
6                         Dihydropterin deaminase
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTable</span>(fit,<span class="dt">coef=</span><span class="dv">3</span>,<span class="dt">sort.by=</span><span class="st">&quot;p&quot;</span>)</code></pre></div>
<pre class="output"><code>                FLYBASE ENTREZID    SYMBOL
FBgn0025111 FBgn0025111    32008      Ant2
FBgn0003360 FBgn0003360    32007      sesB
FBgn0026562 FBgn0026562    43230     SPARC
FBgn0029167 FBgn0029167    39529       Hml
FBgn0035085 FBgn0035085    37991    CG3770
FBgn0039155 FBgn0039155    42865      Kal1
FBgn0040091 FBgn0040091    37590   Ugt58Fa
FBgn0023479 FBgn0023479    39048   Tequila
FBgn0029896 FBgn0029896    31612    CG3168
FBgn0027279 FBgn0027279    33137 l(1)G0196
                                                 GENENAME     logFC
FBgn0025111              Adenine nucleotide translocase 2 -2.863110
FBgn0003360                            stress-sensitive B  3.089242
FBgn0026562       Secreted protein, acidic, cysteine-rich  2.422219
FBgn0029167                                    Hemolectin  2.247086
FBgn0035085 CG3770 gene product from transcript CG3770-RA  2.677008
FBgn0039155                  Kallmann syndrome 1 ortholog  4.272254
FBgn0040091 CG4414 gene product from transcript CG4414-RB  1.559497
FBgn0023479 CG4821 gene product from transcript CG4821-RG  1.576335
FBgn0029896 CG3168 gene product from transcript CG3168-RG  2.312097
FBgn0027279                              lethal (1) G0196  1.180272
              AveExpr         t      P.Value    adj.P.Val        B
FBgn0025111  6.253448 -31.24326 9.756044e-13 4.734428e-09 19.04663
FBgn0003360  7.778378  30.57062 1.258487e-12 4.734428e-09 19.04370
FBgn0026562 11.359837  24.25592 1.868075e-11 4.685133e-08 16.57869
FBgn0029167  7.825170  21.77945 6.509505e-11 1.224438e-07 15.57725
FBgn0035085  5.180607  18.97509 3.188148e-10 4.797524e-07 13.66077
FBgn0039155  4.800907  17.42222 8.476440e-10 1.062946e-06 12.15024
FBgn0040091  6.245766  15.43293 3.363833e-09 3.615639e-06 11.72125
FBgn0023479  7.768886  15.06473 4.419245e-09 3.877495e-06 11.43434
FBgn0029896  4.881577  15.00033 4.638152e-09 3.877495e-06 11.25267
FBgn0027279  7.760889  14.37303 7.503535e-09 5.645660e-06 10.89435
</code></pre>
<p>Let’s check the expression of pasilla.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ps &lt;-<span class="st"> </span><span class="kw">grep</span>(<span class="st">&quot;pasilla&quot;</span>,fit$genes$GENENAME)
<span class="kw">topTable</span>(fit[ps,],<span class="dt">coef=</span><span class="dv">3</span>)</code></pre></div>
<pre class="output"><code>                FLYBASE ENTREZID SYMBOL GENENAME    logFC  AveExpr       t
FBgn0261552 FBgn0261552    44258     ps  pasilla 1.845368 8.344996 11.5651
                P.Value   adj.P.Val        B
FBgn0261552 8.37355e-08 8.37355e-08 8.425095
</code></pre>
<h2 id="plots-after-testing-for-de">Plots after testing for DE</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plotMD</span>(fit,<span class="dt">coef=</span><span class="dv">3</span>,<span class="dt">status=</span>results[,<span class="st">&quot;UVsT&quot;</span>])

<span class="kw">volcanoplot</span>(fit,<span class="dt">coef=</span><span class="dv">3</span>,<span class="dt">highlight=</span><span class="dv">100</span>,<span class="dt">names=</span>fit$genes$symbol)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-35-1.png" title="plot of chunk unnamed-chunk-35" alt="plot of chunk unnamed-chunk-35" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">stripchart</span>(v$E[<span class="st">&quot;FBgn0025111&quot;</span>,]~targets$Group)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-36-1.png" title="plot of chunk unnamed-chunk-36" alt="plot of chunk unnamed-chunk-36" style="display: block; margin: auto;" /></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># Check expression of Pasilla</span>
<span class="kw">stripchart</span>(v$E[<span class="st">&quot;FBgn0261552&quot;</span>,]~targets$Group)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-37-1.png" title="plot of chunk unnamed-chunk-37" alt="plot of chunk unnamed-chunk-37" style="display: block; margin: auto;" /></p>
<h2 id="testing-relative-to-a-threshold">Testing relative to a threshold</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">fit.treat &lt;-<span class="st"> </span><span class="kw">treat</span>(fit,<span class="dt">lfc=</span><span class="dv">1</span>)
res.treat &lt;-<span class="st"> </span><span class="kw">decideTests</span>(fit.treat)
<span class="kw">summary</span>(res.treat)</code></pre></div>
<pre class="output"><code>    Int SEvsPE UVsT
-1    2     13   10
0   350   7510 7496
1  7172      1   18
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">topTreat</span>(fit.treat,<span class="dt">coef=</span><span class="dv">3</span>)</code></pre></div>
<pre class="output"><code>                FLYBASE ENTREZID SYMBOL
FBgn0003360 FBgn0003360    32007   sesB
FBgn0025111 FBgn0025111    32008   Ant2
FBgn0026562 FBgn0026562    43230  SPARC
FBgn0039155 FBgn0039155    42865   Kal1
FBgn0029167 FBgn0029167    39529    Hml
FBgn0035085 FBgn0035085    37991 CG3770
FBgn0034736 FBgn0034736    37572 CG6018
FBgn0000071 FBgn0000071    40831    Ama
FBgn0029896 FBgn0029896    31612 CG3168
FBgn0039827 FBgn0039827    43689 CG1544
                                                 GENENAME     logFC
FBgn0003360                            stress-sensitive B  3.089242
FBgn0025111              Adenine nucleotide translocase 2 -2.863110
FBgn0026562       Secreted protein, acidic, cysteine-rich  2.422219
FBgn0039155                  Kallmann syndrome 1 ortholog  4.272254
FBgn0029167                                    Hemolectin  2.247086
FBgn0035085 CG3770 gene product from transcript CG3770-RA  2.677008
FBgn0034736 CG6018 gene product from transcript CG6018-RA  3.407306
FBgn0000071                                       Amalgam -2.572434
FBgn0029896 CG3168 gene product from transcript CG3168-RG  2.312097
FBgn0039827 CG1544 gene product from transcript CG1544-RA  3.925533
              AveExpr          t      P.Value    adj.P.Val
FBgn0003360  7.778378  20.674784 5.940461e-11 2.711163e-07
FBgn0025111  6.253448 -20.330911 7.206708e-11 2.711163e-07
FBgn0026562 11.359837  14.241997 4.158446e-09 1.042938e-05
FBgn0039155  4.800907  13.344224 8.653631e-09 1.627748e-05
FBgn0029167  7.825170  12.087142 2.578989e-08 3.880863e-05
FBgn0035085  5.180607  11.886920 3.099362e-08 3.886600e-05
FBgn0034736  3.372966   8.780952 7.931358e-07 8.094637e-04
FBgn0000071  4.209333  -8.710932 8.606738e-07 8.094637e-04
FBgn0029896  4.881577   8.512571 1.091548e-06 8.898178e-04
FBgn0039827  3.204658   8.449359 1.182639e-06 8.898178e-04
</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plotMD</span>(fit.treat,<span class="dt">coef=</span><span class="dv">3</span>,<span class="dt">status=</span>res.treat[,<span class="st">&quot;UVsT&quot;</span>])
<span class="kw">abline</span>(<span class="dt">h=</span><span class="dv">0</span>,<span class="dt">col=</span><span class="st">&quot;grey&quot;</span>)</code></pre></div>
<p><img src="fig-09/unnamed-chunk-39-1.png" title="plot of chunk unnamed-chunk-39" alt="plot of chunk unnamed-chunk-39" style="display: block; margin: auto;" /></p>
<h2 id="gene-set-testing-with-goana">Gene set testing with goana</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">go &lt;-<span class="st"> </span><span class="kw">goana</span>(fit, <span class="dt">coef=</span><span class="st">&quot;UVsT&quot;</span>,<span class="dt">species =</span> <span class="st">&quot;Dm&quot;</span>, <span class="dt">geneid=</span><span class="st">&quot;ENTREZID&quot;</span>)
<span class="kw">topGO</span>(go, <span class="dt">n=</span><span class="dv">10</span>)</code></pre></div>
<pre class="output"><code>                                                     Term Ont   N Up Down
GO:0019991                      septate junction assembly  BP  27  0   12
GO:0044724 single-organism carbohydrate catabolic process  BP  22  1   10
GO:0043297                       apical junction assembly  BP  34  0   12
GO:0016052                 carbohydrate catabolic process  BP  23  1   10
GO:0060856           establishment of blood-brain barrier  BP  14  0    8
GO:0044421                      extracellular region part  CC 154 22   27
GO:0005886                                plasma membrane  CC 496 36   58
GO:0007043                    cell-cell junction assembly  BP  35  0   12
GO:0005576                           extracellular region  CC 269 35   38
GO:0071944                                 cell periphery  CC 577 43   64
                   P.Up       P.Down
GO:0019991 1.000000e+00 1.460559e-08
GO:0044724 7.321697e-01 1.877361e-07
GO:0043297 1.000000e+00 3.129742e-07
GO:0016052 7.477650e-01 3.144861e-07
GO:0060856 1.000000e+00 3.503620e-07
GO:0044421 6.961118e-05 3.604688e-07
GO:0005886 9.246429e-02 4.168760e-07
GO:0007043 1.000000e+00 4.506965e-07
GO:0005576 4.470930e-06 5.296478e-07
GO:0071944 4.980817e-02 6.646739e-07
</code></pre>
        </div>
      </div>
      </article>
      
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
