---
layout: page
title: RNAseq analysis in R
---

In this workshop, you will be learning how to analyse RNA-seq count data, using R. This will include reading the data into R, quality control and performing differential expression analysis and gene set testing, with a focus on the limma-voom analysis workflow. You will learn how to generate common plots for analysis and visualisation of gene expression data, such as boxplots and heatmaps. You will also be learning how alignment and counting of raw RNA-seq data can be performed in R. In addition, you will be introduced to Degust, a user-friendly interactive tool for RNA-seq analysis. This workshop is aimed at biologists interested in learning how to perform differential expression analysis of RNA-seq data when reference genomes are available.

> ## Prerequisites {.prereq}
>
> Some basic R knowledge is assumed.
> If you are not familiar with the R statistical programming language we
> strongly encourage you to work through an introductory R course before
> attempting these materials.
> We recommend the Software Carpentry
> [R for Reproducible Scientific Analysis](http://swcarpentry.github.io/r-novice-gapminder/)
> lessons up to and including vectorisation.
>

## Topics

### R for RNAseq

- [Project management with RStudio](01-rstudio-intro.html)
- [Seeking help](02-seeking-help.html)
- [Data structures](03-data-structures-part1.html)
- [Data frames and reading in data](04-data-structures-part2.html)
- [Subsetting data](05-data-subsetting.html)

### RNAseq

- [Differential expression](06-rnaseq-day1.html)
- [Alignment and feature counting](07-rnaseq-day2.html)

- Gene set testing
- Applying RNAseq

## Other Resources
