<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8">
    <meta name="generator" content="pandoc">
    <title>RNAseq analysis in R</title>
    <link rel="shortcut icon" type="image/x-icon" href="/favicon.ico" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap.css" />
    <link rel="stylesheet" type="text/css" href="css/bootstrap/bootstrap-theme.css" />
    <link rel="stylesheet" type="text/css" href="css/swc.css" />
    <link rel="alternate" type="application/rss+xml" title="Software Carpentry Blog" href="http://software-carpentry.org/feed.xml"/>
    <meta charset="UTF-8" />
    <!-- HTML5 shim, for IE6-8 support of HTML5 elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body class="lesson">
    <div class="container card">
      
      <article>
      <div class="row">
        <div class="col-md-10 col-md-offset-1">
                    <h1 class="title">RNAseq analysis in R</h1>
          <p>In this workshop, you will be learning how to analyse RNA-seq count data, using R. This will include reading the data into R, quality control and performing differential expression analysis and gene set testing, with a focus on the limma-voom analysis workflow. You will learn how to generate common plots for analysis and visualisation of gene expression data, such as boxplots and heatmaps. You will also be learning how alignment and counting of raw RNA-seq data can be performed in R. In addition, you will be introduced to Degust, a user-friendly interactive tool for RNA-seq analysis. This workshop is aimed at biologists interested in learning how to perform differential expression analysis of RNA-seq data when reference genomes are available.</p>
<section class="prereq panel panel-warning">
<div class="panel-heading">
<h2 id="prerequisites"><span class="glyphicon glyphicon-education"></span>Prerequisites</h2>
</div>
<div class="panel-body">
<p>Some basic R knowledge is assumed. If you are not familiar with the R statistical programming language we strongly encourage you to work through an introductory R course before attempting these materials. We recommend the Software Carpentry <a href="http://swcarpentry.github.io/r-novice-gapminder/">R for Reproducible Scientific Analysis</a> lessons up to and including vectorisation.</p>
</div>
</section>
<h2 id="data">Data</h2>
<ul>
<li>Day 1: <a href="https://figshare.com/s/1d788fd384d33e913a2a" class="uri">https://figshare.com/s/1d788fd384d33e913a2a</a></li>
<li>Day 2: <a href="https://figshare.com/s/f5d63d8c265a05618137" class="uri">https://figshare.com/s/f5d63d8c265a05618137</a></li>
</ul>
<p>The Day 2 data is ~400MB, so we recommend you download this somewhere with a decent internet connection e.g.Â at home.</p>
<h2 id="topics">Topics</h2>
<h3 id="r-for-rnaseq">R for RNAseq</h3>
<p>Note, this topic is optional for those with substantial R programming experience. However, make sure you complete the challenges in topic 1 to ensure you have set up your RStudio environment correctly and have downloaded the data required for topics 6-8.</p>
<ol style="list-style-type: decimal">
<li><a href="01-rstudio-intro.html">Project management with RStudio</a></li>
<li><a href="02-seeking-help.html">Seeking help</a></li>
<li><a href="03-data-structures-part1.html">Data structures</a></li>
<li><a href="04-data-structures-part2.html">Data frames and reading in data</a></li>
<li><a href="05-data-subsetting.html">Subsetting data</a></li>
</ol>
<h3 id="rnaseq-anlaysis-in-r">RNAseq anlaysis in R</h3>
<ol start="6" style="list-style-type: decimal">
<li><a href="06-rnaseq-day1.html">Differential expression and gene set testing</a></li>
<li><a href="07-rnaseq-day2.html">Alignment and feature counting</a></li>
<li><a href="08-applying-rnaseq.html">Applying RNAseq</a> (<a href="09-applying-rnaseq-solutions.html">solutions</a>)</li>
</ol>
<div class="references">

</div>
        </div>
      </div>
      </article>
      
    </div>
    <!-- Javascript placed at the end of the document so the pages load faster -->
    <script src="http://software-carpentry.org/v5/js/jquery-1.9.1.min.js"></script>
    <script src="css/bootstrap/bootstrap-js/bootstrap.js"></script>
    <script src='https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>
  </body>
</html>
